twine/commands/upload.py

```diff
@@ -48,6 +53,39 @@ def skip_upload(response, skip_existing, package):
             (response.status_code == 403 and msg_403 in response.text)))
 
 
+def check_status_code(response, verbose):
+    """
+    Additional safety net to catch response code 410 in case the
+    UploadToDeprecatedPyPIDetected exception breaks.
+    Also includes a check for response code 405 and prints helpful error
+    message guiding users to the right repository endpoints.
+    """
+    if (response.status_code == 410 and
+            response.url.startswith(("https://pypi.python.org",
+                                     "https://testpypi.python.org"))):
+        print("It appears you're uploading to pypi.python.org (or "
+              "testpypi.python.org). You've received a 410 error response. "
+              "Uploading to those sites is deprecated. The new sites are "
+              "pypi.org and test.pypi.org. Try using "
+              "https://upload.pypi.org/legacy/ "
+              "(or https://test.pypi.org/legacy/) to upload your packages "
+              "instead. These are the default URLs for Twine now. More at "
+              "https://packaging.python.org/guides/migrating-to-pypi-org/ ")
+    elif response.status_code == 405 and "pypi.org" in response.url:
+        print(f"You probably want one of these two URLs: {DEFAULT_REPOSITORY} "
```

Comment:
Do you have an example of the complete command output when this message is printed? It seems like we might want a little more information to indicate that an incorrect repository URL was configured (akin to the message above), but maybe it's already obvious.

Reply:
Here you go:

```
(env) ➜  adityasaky-test-package twine upload --repository-url https://test.pypi.org/ dist/*
Enter your username: adityasaky
Enter your password: 
Uploading distributions to https://test.pypi.org/
Uploading adityasaky_test_package-0.0.1-py3-none-any.whl
100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 16.7k/16.7k [00:00<00:00, 116kB/s]
You probably want one of these two URLs: https://upload.pypi.org/legacy/ or https://test.pypi.org/legacy/.
NOTE: Try --verbose to see response content.
HTTPError: 405 Client Error: Method Not Allowed for url: https://test.pypi.org/
```

With `--verbose`:

```
(env) ➜  adityasaky-test-package twine upload --verbose --repository-url https://test.pypi.org/ dist/*
Enter your username: adityasaky
Enter your password: 
Uploading distributions to https://test.pypi.org/
Uploading adityasaky_test_package-0.0.1-py3-none-any.whl
100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 16.7k/16.7k [00:00<00:00, 108kB/s]
You probably want one of these two URLs: https://upload.pypi.org/legacy/ or https://test.pypi.org/legacy/.
Content received from server:
<html>
 <head>
  <title>405 Method Not Allowed</title>
 </head>
 <body>
  <h1>405 Method Not Allowed</h1>
  The method POST is not allowed for this resource. <br/><br/>

 </body>
</html>
HTTPError: 405 Client Error: Method Not Allowed for url: https://test.pypi.org/
```

Reply:
Added a line in `a31ba7b` at the end of the current message: `Check your --repository-url value.`.

Reply:
Thanks for this. Seeing it makes me wonder about putting the helpful message after the error (also for the 410 response, which could be shortened). Looking at the code, I think that might be something like moving it to the `except` block.

I also note that we raise [custom exceptions](https://github.com/pypa/twine/blob/master/twine/exceptions.py) in other places. Maybe this code could be re-worked to raise `UploadToDeprecatedPyPIDetected` or a new `PyPIMethodNotAllowed` if we recognize those conditions, otherwise we re-raise the `HTTPError` as it does now?

Reply:
Sure, that makes sense too, and I considered that, but I wasn't sure if you _wanted_ to raise `UploadToDeprecatedPyPIDetected` as that's already being raised elsewhere first, and the comment spoke about this being around in case that check failed.

Adding a new exception for error 405 also makes sense.